### DISCUSSION TITLE
sourceUrl of a published package not up-to-date

### DISCUSSION THREAD

**Discussion by KnisterPeter** [renovatebot/renovate#13940] [Request Help] [answered]
Hi,

we are using nexus as our internal npm registry. We have a package which switch the sourceUrl from a repository into a mono-repository. The repository metadata is updated in our `package.json` file and a new release was published with the `latest` dist-tag.
So far okay. Not we recognized that renovate does not group that package with the other ones from our monorepo and traced the problem to an out-dated `sourceUrl`.

This might be an issue with nexus, since in the nexus package data, the repository key is indeed outdated. But there is a repository key per version which is up-to date.
We looked into the source of renovate and found that the global repository is taken and only if that is falsy the one from the `latest` dist-tag is choosen:

https://github.com/renovatebot/renovate/blob/790ff0196f75900da4b80161e0990b32d3406a39/lib/datasource/npm/get.ts#L101

What is the reason to this implementation behavior? Isn't the latest versions information the one to prefer?

**Comment by rarkins** [comment, marked as answer]
I think we should do this (and thought we already did, but my memory may be wrong):

- package sourceUrl is as above (defaults to what's in the package.json response, or latest release repository if missing)
- each "release" in our response should have sourceUrl populated if it has a URL which differs from the package sourceUrl

That way you'd get the behavior you are looking for, and we could support e.g. older versions with one sourceUrl and newer ones with a different sourceUrl

**Reply by KnisterPeter** [reply]
Should I open an issue for this? Because the current behavior is different from what you describe (if I understood you).

**Reply by KnisterPeter** [reply]
I've also opened an issue at nexus to match npm public repo behavior:
https://issues.sonatype.org/browse/NEXUS-31073

**Reply by rarkins** [reply]
Yes, please open an issue, maybe just npm specific and you can call it "Per-release sourceUrl support for npm datasource"

**Reply by KnisterPeter** [reply]
Thank you :+1: 
https://github.com/renovatebot/renovate/issues/13943
